x/feerefunder/keeper/keeper_test.go

```diff
@@ -156,18 +162,113 @@ func TestKeeperLockFees(t *testing.T) {
+	require.NoError(t, err)
+	require.Equal(t, p, types.PayerInfo{
+		Sender:   sdk.MustAccAddressFromBech32(TestAddress),
+		FeePayer: sdk.MustAccAddressFromBech32(TestAddress),
+	})
+}
+
+func TestKeeperLockFeesAtFeePayeer(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	bankKeeper := mock_types.NewMockBankKeeper(ctrl)
+	channelKeeper := mock_types.NewMockChannelKeeper(ctrl)
+	feegrantKeeper := mock_types.NewMockFeeGrantKeeper(ctrl)
+	feeAllowance := mock_feegrant.NewMockFeeAllowanceI(ctrl)
+	k, ctx := testutil_keeper.FeeKeeper(t, channelKeeper, bankKeeper, feegrantKeeper)
+
+	payerInfo := types.PayerInfo{
+		Sender:   sdk.MustAccAddressFromBech32(testutil.TestOwnerAddress),
+		FeePayer: sdk.MustAccAddressFromBech32(TestAddress),
+	}
+
+	k.SetParams(ctx, types.Params{
+		MinFee: types.Fee{
+			RecvFee:    nil,
+			AckFee:     sdk.NewCoins(sdk.NewCoin("denom1", sdk.NewInt(100)), sdk.NewCoin("denom2", sdk.NewInt(100))),
+			TimeoutFee: sdk.NewCoins(sdk.NewCoin("denom1", sdk.NewInt(100)), sdk.NewCoin("denom2", sdk.NewInt(100))),
+		},
+	})
+
+	packet := types.PacketID{
+		ChannelId: "channel-0",
+		PortId:    "transfer",
+		Sequence:  111,
+	}
+	ackFee := sdk.NewCoins(sdk.NewCoin("denom1", sdk.NewInt(101)))
+	timeoutFee := sdk.NewCoins(sdk.NewCoin("denom1", sdk.NewInt(101)))
+	validFee := types.Fee{
+		RecvFee:    nil,
+		AckFee:     ackFee,
+		TimeoutFee: timeoutFee,
+	}
+
+	channelKeeper.EXPECT().GetChannel(ctx, packet.PortId, packet.ChannelId).Return(channeltypes.Channel{}, true)
+	feegrantKeeper.EXPECT().GetAllowance(ctx, payerInfo.FeePayer, payerInfo.Sender).Return(nil, fmt.Errorf("feegrant error"))
+	err := k.LockFees(ctx, payerInfo, packet, validFee)
+	require.ErrorContains(t, err, "feegrant error")
+
+	channelKeeper.EXPECT().GetChannel(ctx, packet.PortId, packet.ChannelId).Return(channeltypes.Channel{}, true)
+	fees := append(ackFee, timeoutFee...)
+	feeAllowance.EXPECT().Accept(ctx, fees, []sdk.Msg{}).Return(false, fmt.Errorf("fee allowance accept error"))
+	feegrantKeeper.EXPECT().GetAllowance(ctx, payerInfo.FeePayer, payerInfo.Sender).Return(feeAllowance, nil)
+	err = k.LockFees(ctx, payerInfo, packet, validFee)
+	require.ErrorContains(t, err, "fee allowance accept error")
+
+	channelKeeper.EXPECT().GetChannel(ctx, packet.PortId, packet.ChannelId).Return(channeltypes.Channel{}, true)
+	feeAllowance.EXPECT().Accept(ctx, fees, []sdk.Msg{}).Return(false, nil)
+	feegrantKeeper.EXPECT().GetAllowance(ctx, payerInfo.FeePayer, payerInfo.Sender).Return(feeAllowance, nil)
+	bankKeeper.EXPECT().SendCoinsFromAccountToModule(ctx, payerInfo.FeePayer, types.ModuleName, validFee.Total()).Return(nil)
+	err = k.LockFees(ctx, payerInfo, packet, validFee)
 	require.NoError(t, err)
 	require.Equal(t, sdk.Events{
 		sdk.NewEvent(
 			types.EventTypeLockFees,
-			sdk.NewAttribute(types.AttributeKeyPayer, payer.String()),
+			sdk.NewAttribute(types.AttributeKeyPayer, payerInfo.Sender.String()),
```

Comment:
bc one is actually tx sender and the other is actually fee payer and both of them are kind of payment information. So mb you have any good idea for a struct name which consist of sender and fee payer?

Reply:
no no naming is good. I meant why the `types.AttributeKeyPayer` attribute value here is `payerInfo.Sender` if we've passed a payerInfo containing the `FeePayer` field into the `LockFees`? we use the `payer` as the `types.AttributeKeyPayer` value which is set to `payerInfo.FeePayer` [here](https://github.com/neutron-org/neutron/blob/efff248f4d848ad1329c9e98aa0419de521c66a0/x/feerefunder/keeper/keeper.go#L87). how then it can be equal `payerInfo.Sender` here, not `payerInfo.FeePayer`?

Reply:
oh, got it. so the payer is Sender when we don't have FeePayer. 
<img width="903" alt="image" src="https://user-images.githubusercontent.com/198887/214836086-7a441789-8623-45c9-b270-e4b754cd9efa.png">

Reply:
thanks, nice spot 👍